swarm/network/simulations/discovery/dynamic_test.go

```diff
@@ -0,0 +1,538 @@
+		go func() {
+			for _, n := range ids {
+				if err := net.Start(n); err != nil {
+					t.Fatalf("error starting node: %s", err)
+				}
+				log.Info("network start returned", "node", n)
+			}
+		}()
+		return nil
+
+	}
+
+	check := func(ctx context.Context, nodeId discover.NodeID) (bool, error) {
+		select {
+		case <-ctx.Done():
+			return false, ctx.Err()
+		default:
+		}
+		log.Info("trigger expect up", "node", nodeId)
+		return true, nil
+	}
+
+	timeout := 30 * time.Second
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+	result := simulations.NewSimulation(net).Run(ctx, &simulations.Step{
+		Action:  action,
+		Trigger: trigger,
+		Expect: &simulations.Expectation{
+			Nodes: ids,
+			Check: check,
+		},
+	})
+
+	if result.Error != nil {
+		t.Fatal(result.Error)
+	}
+	close(*quitC)
+
+	// sim step 2
+	// connect the three bootnodes together
+	// connect each of the other nodes to a random bootnode
+	q = make(chan struct{})
+	quitC = &q
+	action = func(ctx context.Context) error {
+		go func(quitC chan struct{}) {
+			for {
+				select {
+				case ev := <-events:
+					if ev.Type == simulations.EventTypeConn {
+						if ev.Conn.Up {
+							log.Info(fmt.Sprintf("got conn up event %v", ev))
+							trigger <- ev.Conn.One
+						}
+					}
+				case <-quitC:
+					log.Debug("got quit action connect")
+					return
+				}
+			}
+		}(*quitC)
+		go func() {
+			for i := range ids {
+				var j int
+				if i == 0 {
+					continue
+				}
+				if i < len(bootNodes) {
+					j = i - 1
+				} else {
+					j = rand.Intn(len(bootNodes) - 1)
+				}
+
+				if err := net.Connect(ids[i], ids[j]); err != nil {
+					t.Fatalf("error connecting node %x => bootnode %x: %s", ids[i], ids[j], err)
+				}
+				log.Info("network connect returned", "one", ids[i], "other", ids[j])
+			}
+		}()
+		return nil
+	}
+
+	check = func(ctx context.Context, id discover.NodeID) (bool, error) {
+		select {
+		case <-ctx.Done():
+			return false, ctx.Err()
+		default:
+		}
+		log.Info("trigger expect conn", "node", id)
+		return true, nil
+	}
+
+	timeout = 30 * time.Second
+	ctx, cancel = context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+	result = simulations.NewSimulation(net).Run(ctx, &simulations.Step{
+		Action:  action,
+		Trigger: trigger,
+		Expect: &simulations.Expectation{
+			Nodes: ids[1:],
+			Check: check,
+		},
+	})
+	if result.Error != nil {
+		t.Fatal(result.Error)
+	}
+	close(*quitC)
+	*quitC = nil
+	sub.Unsubscribe()
+	sub = nil
+
+	// check health of all nodes
+	ctrl := newNodeCtrl(net, ids)
+	for _, n := range net.GetNodes() {
+		err := ctrl.checkHealth(n, 0)
+		if err != nil {
+			t.Fatalf("Node %v failed health check", n)
+		}
+		log.Info("node health ok", "node", n)
+	}
+
+	// asynchronously restart random nodes
+	wg := sync.WaitGroup{}
+	for i := 0; i < int(numUpDowns); i++ {
+		wg.Add(1)
+		dur := ctrl.randomDelay(0)
+		time.Sleep(dur)
+		go func() {
+			if err := ctrl.nodeUpDown(); err != nil {
```

Comment:
in this scenario, I understand that only the current node will be checked for `heatlhy` during `nodeUpDown()`; but a node disconnecting could also affect other nodes; shouldn't ALL nodes be checked for healthy after restarting  a single node?

Reply:
Hmm, that sounds like a different test.